File: Java/Android/Canvas/src/org/loon/framework/android/game/action/sprite/effect/FadeEffect.java
package org.loon.framework.android.game.action.sprite.effect;

import org.loon.framework.android.game.action.sprite.ISprite;
import org.loon.framework.android.game.core.LObject;
import org.loon.framework.android.game.core.LSystem;
import org.loon.framework.android.game.core.geom.RectBox;
import org.loon.framework.android.game.core.graphics.LColor;
import org.loon.framework.android.game.core.graphics.device.LGraphics;

import android.graphics.Bitmap;

/**
 * Copyright 2008 - 2010
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loonframework
 * @author chenpeng
 * @email：ceponline@yahoo.com.cn
 * @version 0.1
 */
public class FadeEffect extends LObject implements ISprite {
	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;

	public LColor color;

	public float time;

	public float currentFrame;

	public int type;

	public boolean stop;

	private float opacity;

	private int offsetX, offsetY;

	private int width;

	private int height;

	private boolean visible;

	public static FadeEffect getInstance(int type, LColor c) {
		return new FadeEffect(c, 60, type, LSystem.getSystemHandler().getWidth(),
				LSystem.getSystemHandler().getHeight());

	}

	public FadeEffect(LColor c, int delay, int type, int w, int h) {
		this.visible = true;
		this.type = type;
		this.setDelay(delay);
		this.setColor(c);
		this.width = w;
		this.height = h;
	}

	public float getDelay() {
		return time;
	}

	public void setDelay(int delay) {
		this.time = delay;
		if (type == TYPE_FADE_IN) {
			this.currentFrame = this.time;
		} else {
			this.currentFrame = 0;
		}
	}

	public LColor getColor() {
		return color;
	}

	public void setColor(LColor color) {
		this.color = color;
	}

	public float getCurrentFrame() {
		return currentFrame;
	}

	public void setCurrentFrame(float currentFrame) {
		this.currentFrame = currentFrame;
	}

	public boolean isStop() {
		return stop;
	}

	public void setStop(boolean stop) {
		this.stop = stop;
	}

	public int getType() {
		return type;
	}

	public void setType(int type) {
		this.type = type;
	}

	public void setVisible(boolean visible) {
		this.opacity = visible ? 255 : 0;
		this.visible = visible;
	}

	public boolean isVisible() {
		return visible;
	}

	public void setOpacity(float opacity) {
		this.opacity = opacity;
	}

	public float getOpacity() {
		return opacity;
	}

	public void createUI(LGraphics g) {
		if (!visible) {
			return;
		}
		if (stop) {
			return;
		}
		float op = (currentFrame / time) * 255;
		setOpacity(op);
		if (opacity > 0) {
			LColor tempColor = g.getColor();
			g.setColor(color.getRed(), color.getGreen(), color.getBlue(),
					(int) opacity);
			g.fillRect(offsetX + this.x(), offsetY + this.y(), width, height);
			g.setColor(tempColor);
			return;
		}
	}

	public void update(long timer) {
		if (type == TYPE_FADE_IN) {
			currentFrame--;
			if (currentFrame == 0) {
				setOpacity(0);
				stop = true;
			}
		} else {
			currentFrame++;
			if (currentFrame == time) {
				setOpacity(0);
				stop = true;
			}
		}
	}

	public float getAlpha() {
		return 0;
	}

	public RectBox getCollisionBox() {
		return getRect(x(), y(), getWidth(), getHeight());
	}

	public int getHeight() {
		return height;
	}

	public int getWidth() {
		return width;
	}

	public int getOffsetX() {
		return offsetX;
	}

	public void setOffsetX(int offsetX) {
		this.offsetX = offsetX;
	}

	public int getOffsetY() {
		return offsetY;
	}

	public void setOffsetY(int offsetY) {
		this.offsetY = offsetY;
	}

	public Bitmap getBitmap() {
		return null;
	}

	public void dispose() {

	}

}

File: C#/Silverlight/source/LSilverlight-0.3.3/Loon.Action.Sprite.Effect/FadeEffect.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Loon.Core;
using Loon.Core.Graphics;
using Loon.Core.Graphics.OpenGL;
using Microsoft.Xna.Framework;
using Loon.Core.Geom;

namespace Loon.Action.Sprite.Effect
{
    public class FadeEffect : LObject, ISprite
    {
        
	    public const int TYPE_FADE_IN = 0;

        public const int TYPE_FADE_OUT = 1;

        public LColor color;

        public float time;

        public float currentFrame;

        public int type;

        public bool stop;

        private float opacity;

        private int offsetX, offsetY;

        private int width;

        private int height;

        private bool visible;

        public static FadeEffect GetInstance(int type, LColor c)
        {
            return new FadeEffect(c, 120, type, LSystem.screenRect.width,
                    LSystem.screenRect.height);

        }

        public FadeEffect(LColor c, int delay, int type, int w, int h)
        {
            this.visible = true;
            this.type = type;
            this.SetDelay(delay);
            this.SetColor(c);
            this.width = w;
            this.height = h;
        }

        public float GetDelay()
        {
            return time;
        }

        public void SetDelay(int delay)
        {
            this.time = delay;
            if (type == TYPE_FADE_IN)
            {
                this.currentFrame = this.time;
            }
            else
            {
                this.currentFrame = 0;
            }
        }

        public LColor GetColor()
        {
            return color;
        }

        public void SetColor(LColor color)
        {
            this.color = color;
        }

        public float GetCurrentFrame()
        {
            return currentFrame;
        }

        public void SetCurrentFrame(float currentFrame)
        {
            this.currentFrame = currentFrame;
        }

        public bool IsStop()
        {
            return stop;
        }

        public void SetStop(bool stop)
        {
            this.stop = stop;
        }

        public int GetCode()
        {
            return type;
        }

        public void SetCode(int type)
        {
            this.type = type;
        }

        public void SetVisible(bool visible)
        {
            this.opacity = visible ? 255 : 0;
            this.visible = visible;
        }

        public bool IsVisible()
        {
            return visible;
        }

        public void SetOpacity(float opacity)
        {
            this.opacity = opacity;
        }

        public float GetOpacity()
        {
            return opacity;
        }

        public void CreateUI(GLEx g)
        {
            if (!visible)
            {
                return;
            }
            if (stop)
            {
                return;
            }
            float op = (currentFrame / time);
            SetOpacity(op);
            if (opacity > 0)
            {
                byte a = (byte)(this.opacity * 255);
                g.FillRect(offsetX + this.X(), offsetY + this.Y(), width, height, new Color(color.R, color.G, color.B, a));
                return;
            }
        }

        public override void Update(long timer)
        {
            if (type == TYPE_FADE_IN)
            {
                currentFrame--;
                if (currentFrame == 0)
                {
                    SetOpacity(0);
                    stop = true;
                }
            }
            else
            {
                currentFrame++;
                if (currentFrame == time)
                {
                    SetOpacity(0);
                    stop = true;
                }
            }
        }

        public RectBox GetCollisionBox()
        {
            return GetRect(X(), Y(),GetWidth(), GetHeight());
        }

        public override int GetHeight()
        {
            return height;
        }

        public override int GetWidth()
        {
            return width;
        }

        public int GetOffsetX()
        {
            return offsetX;
        }

        public void SetOffsetX(int offsetX)
        {
            this.offsetX = offsetX;
        }

        public int GetOffsetY()
        {
            return offsetY;
        }

        public void SetOffsetY(int offsetY)
        {
            this.offsetY = offsetY;
        }

        public LTexture GetBitmap()
        {
            return null;
        }

        public void Dispose()
        {

        }

    }
}
